File: jpp/lab4/dining_philosophers.go
// implementation of dining philosophers in go language 

package main

import (
    "fmt"
    "sync"
    "time"
	"os"
	"strconv"
)


type Fork struct{ sync.Mutex }

type Philosopher struct {
    id             int
    leftFork, rightFork *Fork
}

func (p Philosopher) start(wg *sync.WaitGroup, NumMeals int) {
    var lastMeal  = time.Now()
	var currentTime = time.Now()
	var timeElapsed time.Duration

	
	for i := 0; i < NumMeals; i++ {
        fmt.Printf("%d Rozmysla %d\n", p.id + 1, i + 1)
		// czas na rozmyslanie
		time.Sleep(time.Millisecond )

		p.leftFork.Lock()
		fmt.Printf("%d Podnosi lewy \n", p.id + 1)
        p.rightFork.Lock()
		fmt.Printf("%d Podnosi prawy \n", p.id + 1)

		currentTime = time.Now()
		timeElapsed = currentTime.Sub(lastMeal)
		
		for timeElapsed > time.Millisecond * 6 {
			fmt.Printf("%d Umarl z glodu\n", p.id + 1)
			fmt.Printf("Time elapsed: %f \n", timeElapsed.Seconds())
		}

        fmt.Printf("%d Je %d\n", p.id+1,i+1)
        // czas na jedzenie
		time.Sleep(time.Millisecond )
		lastMeal = time.Now()

        p.rightFork.Unlock()
		fmt.Printf("%d Odklada prawy \n", p.id + 1)
        p.leftFork.Unlock()
		fmt.Printf("%d Odklada lewy \n", p.id + 1)

    }
    wg.Done()
}

func main() {
    NumPhilosophers, _ := strconv.Atoi(os.Args[1])
	NumMeals, _ := strconv.Atoi(os.Args[2])

	var wg sync.WaitGroup
    wg.Add(NumPhilosophers)

	
    forks := make([]*Fork, NumPhilosophers)
    for i := 0; i < NumPhilosophers; i++ {
        forks[i] = &Fork{}
    }

    philosophers := make([]*Philosopher, NumPhilosophers)
    for i := 0; i < NumPhilosophers; i++ {
        if i % 2 == 0 {
			philosophers[i] = &Philosopher{
				id: i, 
				leftFork: forks[(i+1)%NumPhilosophers], 
				rightFork: forks[i],
			}
		}else{
			philosophers[i] = &Philosopher{
				id: i, 
				leftFork: forks[i], 
				rightFork: forks[(i+1)%NumPhilosophers],
			}
		}
		
		go philosophers[i].start(&wg, NumMeals)
    }

    wg.Wait()
}

File: jpp/lab4/DiningPhilosophers.java
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;

class Fork {
    private boolean isTaken = false;

    synchronized void pickUp() throws InterruptedException {
        while (isTaken) {
            wait();
        }
        isTaken = true;
    }

    synchronized void putDown() {
        isTaken = false;
        notify();
    }
}


class Philosopher extends Thread {
    private final int id;
    private final Fork leftFork;
    private final Fork rightFork;
    private final int numMeals;
    private final CountDownLatch latch;

    Philosopher(int id, Fork leftFork, Fork rightFork, int numMeals, CountDownLatch latch) {
        this.id = id;
        this.leftFork = leftFork;
        this.rightFork = rightFork;
        this.numMeals = numMeals;
        this.latch = latch;
    }

    public void run() {
        for (int i = 0; i < numMeals; i++) {
            try {
                System.out.printf("%d Rozmysla %d\n", id + 1, i + 1);
                Thread.sleep(1);  // czas na rozmyslanie

                leftFork.pickUp();
                System.out.printf("%d Podnosi lewy \n", id + 1);

                rightFork.pickUp();
                System.out.printf("%d Podnosi prawy \n", id + 1);

                System.out.printf("%d Je %d\n", id + 1, i + 1);

                rightFork.putDown();
                System.out.printf("%d Odklada prawy \n", id + 1);

                leftFork.putDown();
                System.out.printf("%d Odklada lewy \n", id + 1);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        }
        latch.countDown();  
    }
}

public class DiningPhilosophers {
    public static void main(String[] args) {
        int numPhilosophers = Integer.parseInt(args[0]);
        int numMeals = Integer.parseInt(args[1]);

        CountDownLatch latch = new CountDownLatch(numPhilosophers);

        Fork[] forks = new Fork[numPhilosophers];
        for (int i = 0; i < numPhilosophers; i++) {
            forks[i] = new Fork();
        }

        Philosopher[] philosophers = new Philosopher[numPhilosophers];
        for (int i = 0; i < numPhilosophers; i++) {
            if (i % 2 == 0) {
                philosophers[i] = new Philosopher(i, forks[(i + 1) % numPhilosophers], forks[i], numMeals, latch);
            } else {
                philosophers[i] = new Philosopher(i, forks[i], forks[(i + 1) % numPhilosophers], numMeals, latch);
            }
            philosophers[i].start();
        }

        try {
            latch.await();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }
}